drivers: add driver for encx24j600 ethernet chips

kaspar030 opened the pull request:
This adds a driver for encx24j600 ethernet chips, which are connected through SPI.

Waiting for #3683.

kaspar030 commented:
- fixed some travis static-check complaints

kaspar030 commented:
- squashed, rebsaed

kaspar030 commented:
Does anyone but me have such a device? If not, would someone ACK based on code review?

jfischer-phytec-iot commented:
@kaspar030 I have order one module from Bulgaria, if you can wait a few more days ... then I can test it.

kaspar030 commented:
Hm, the driver doesn't have many dependencies, so it should be easy to rebase... So I can wait. :)

kaspar030 commented:
Hm, actually I think it would be good to merge this. #3096 can then be changed to support multiple ethernet devices, and also it would make it easier for Elli to use the driver, and they actually paid me to write it.
@jfischer-phytec-iot, would you ACK based on the code without testing and test once you've got the hardware?

jfischer-phytec-iot commented:
Did not forget you :-). Unfortunately not yet had time and my board sticks in "Verzollung"

jfischer-phytec-iot commented:
@kaspar030 please rebase on master

haukepetersen commented on drivers/encx24j600/encx24j600.c:
```diff
@@ -0,0 +1,424 @@
+
+    enc_spi_transfer(dev, cmd, result, 4);
+
+//    DEBUG("encx24j600_reg_get: got 0x%x from register 0x%x\n", (result[2] | (result[3] << 8)), reg);
```
pls use c style comments

haukepetersen commented on drivers/encx24j600/encx24j600.c:
```diff
@@ -0,0 +1,424 @@
+    gpio_set(dev->cs);
+}
+
+#if 0
```
if not used anyway, throw this out.

haukepetersen commented on drivers/encx24j600/encx24j600.c:
```diff
@@ -0,0 +1,424 @@
+{
+    char cmd[4] = { WCRU, reg, value, value >> 8 };
+
+//    DEBUG("encx24j600_reg_set: setting 0x%x to register 0x%x\n", (cmd[2] | (cmd[3] << 8)), reg);
```
remove

haukepetersen commented on drivers/encx24j600/encx24j600.c:
```diff
@@ -0,0 +1,424 @@
+{
+    char cmd[4] = { BFSU, reg, mask, mask >> 8 };
+
+//    DEBUG("encx24j600_reg_set: setting mask 0x%x to register 0x%x\n", (cmd[2] | (cmd[3] << 8)), reg);
```
remove

haukepetersen commented on drivers/encx24j600/encx24j600.c:
```diff
@@ -0,0 +1,424 @@
+{
+    char cmd[4] = { BFCU, reg, mask, mask >> 8 };
+
+//    DEBUG("encx24j600_reg_set: clearing mask 0x%x to register 0x%x\n", (cmd[2] | (cmd[3] << 8)), reg);
```
same here

haukepetersen commented on drivers/encx24j600/encx24j600.c:
```diff
@@ -0,0 +1,424 @@
+    enc_spi_transfer(dev, cmd, NULL, 4);
+}
+
+#if 0 /* unused */
```
then remove it?!

haukepetersen commented on drivers/encx24j600/encx24j600.c:
```diff
@@ -0,0 +1,424 @@
+    DEBUG("encx24j600: starting initialization...\n");
+
+    /* setup IO */
+    gpio_init_out(dev->cs, GPIO_PULLUP);
```
adapt to changes in GPIO driver: `gpio_init(dev->cs, GPIO_DIR_OUT, ...)`

haukepetersen commented on drivers/encx24j600/encx24j600.c:
```diff
@@ -0,0 +1,424 @@
+    /* setup IO */
+    gpio_init_out(dev->cs, GPIO_PULLUP);
+    gpio_set(dev->cs);
+    gpio_init_int(dev->int_pin, GPIO_PULLUP, GPIO_FALLING, encx24j600_isr, (void*)dev);
```
same here

haukepetersen commented on drivers/encx24j600/encx24j600.c:
```diff
@@ -0,0 +1,424 @@
+
+static inline void enc_spi_transfer(encx24j600_t *dev, char *out, char *in, int len)
+{
+    gpio_clear(dev->cs);
```
you should lock the SPI bus before you use it -> `spi_lock()`. Other what happens if you share the bus between more than one device?

haukepetersen commented on drivers/encx24j600/encx24j600.c:
```diff
@@ -0,0 +1,424 @@
+    char cmd[4] = { RCRU, reg, 0, 0 };
+    char result[4];
+
+    enc_spi_transfer(dev, cmd, result, 4);
```
lock the bus before transfer

OlegHahm replied:
What about this comment?

kaspar030 replied:
The lock has been added inside  enc_spi_transfer()

haukepetersen commented on drivers/encx24j600/encx24j600.c:
```diff
@@ -0,0 +1,424 @@
+
+static void cmd(encx24j600_t *dev, char cmd) {
+    gpio_clear(dev->cs);
+    spi_transfer_byte(dev->spi, cmd, NULL);
```
lock bus before...

haukepetersen commented on drivers/encx24j600/encx24j600.c:
```diff
@@ -0,0 +1,424 @@
+
+static void cmdn(encx24j600_t *dev, uint8_t cmd, char *out, char *in, int len) {
+    gpio_clear(dev->cs);
+    spi_transfer_byte(dev->spi, cmd, NULL);
```
lock bus...

haukepetersen commented on drivers/encx24j600/encx24j600.c:
```diff
@@ -0,0 +1,424 @@
+            vtimer_usleep(ENCX24J600_INIT_DELAY);
+            reg_set(dev, EUDAST, 0x1234);
+            vtimer_usleep(ENCX24J600_INIT_DELAY);
+        } while (reg_get(dev, EUDAST) != 0x1234);
```
this seems dangerous to me in the current state. It would be safer if you test the SPI connection once before you go into the `do` loops - isn't there a device ID or similar fixed value register you can read out after SPI initialization to verify the connection and return an error code if the device does not respond? This is how we handle it in most other (newer) drivers...

kaspar030 replied:
On 06/16/15 03:05, Hauke Petersen wrote:
> this seems dangerous to me in the current state.

procedure is from datasheet. We could some kind of MAXTRIES, though.

kaspar030 replied:
I've added a check for valid SPI device.

haukepetersen commented on drivers/encx24j600/encx24j600.c:
```diff
@@ -0,0 +1,424 @@
+    .isr = _isr,
+};
+
+static inline void lock(encx24j600_t *dev) {
```
actually, it is not quite clear to me, why this device driver level locking is needed. The lower level SPI bus is protected by it's own lock, and the usage model of the network interface drivers so far is that they are only used inside a single thread (e.g. by a single user), so that device driver level locking is not really needed. Seems I am missing something?!

kaspar030 replied:
On 06/16/15 03:08, Hauke Petersen wrote:
> why this device driver level locking is needed.

Think outside the box.

kaspar030 replied:
Sorry for that. I think locking is cheap enough to make sure the driver itself cannot corrupt itself by calling the external interface from multiple threads.

haukepetersen commented on drivers/encx24j600/encx24j600.c:
```diff
@@ -0,0 +1,424 @@
+    cmd(dev, SETTXRTS);
+
+    /* wait for sending to complete */
+    //while ((reg_get(dev, ECON1) & TXRTS));
```
remove?!

kaspar030 replied:
maybe. I'll add a comment.

haukepetersen commented on drivers/encx24j600/encx24j600.c:
```diff
@@ -0,0 +1,424 @@
+    return len;
+}
+
+/*static void encx24j600_handle_linkstate(encx24j600_t *dev, int state)
```
remove?!

haukepetersen commented on drivers/encx24j600/encx24j600.c:
```diff
@@ -0,0 +1,424 @@
+
+    addr[0] = reg_get(dev, MAADR1);
+    addr[1] = reg_get(dev, MAADR2);
+    addr[2] = reg_get(dev, MAADR3);
```
not quite sure, in which byte-order is the MAC address stored (and read)?

kaspar030 replied:
Not sure either. If read like this, the result is the MAC address the device responds to. If I read the bytes as coming from the register into an uint16_t, and then write them to an uint16_t array like in this function, byte-order doesn't matter, right? buf[0] will have the 1st byte, buf[5] the last?

haukepetersen commented on drivers/encx24j600/encx24j600.c:
```diff
@@ -0,0 +1,424 @@
+    /* decrement available packet count */
+    cmd(dev, SETPKTDEC);
+
+//    DEBUG("encx24j600: Received packet. frame_len=%u next=0x%x, newnext=0x%x ERXTAIL=0x%x\n", (unsigned int)hdr.frame_len, (unsigned int) dev->rx_next_ptr, (unsigned int)hdr.rx_next_ptr, (unsigned int)reg_get(dev, ERXTAIL));
```
remove!?

jfischer-phytec-iot replied:
I found the debug message here is very helpful, I have changed a bit. https://github.com/jfischer-phytec-iot/RIOT/commit/b38a4c9c1519e938b23a7b3c48c721fa04567053

haukepetersen commented on drivers/encx24j600/encx24j600.c:
```diff
@@ -0,0 +1,424 @@
+        lock(dev);
+    }
+
+#if 0
```
remove?!

haukepetersen commented on drivers/encx24j600/include/encx24j600_defines.h:
```diff
@@ -0,0 +1,204 @@
+#define RBSEL       0xc8  /* Read Bank Select */
+
+#define SETTXRTS    0xd4  /* Sets TXRTS (ECON1<1>), sends an Ethernet packet */
+
```
newline...

kaspar030 replied:
I'm sorry, I don't share that style preference. If I use newlines to group lines within the doxygen scope, I like that newline at the end of the scope.

kaspar030 replied:
at the same time, I don't care at all. will change.

haukepetersen commented on drivers/encx24j600/include/encx24j600_defines.h:
```diff
@@ -0,0 +1,204 @@
+#define ERXRDPT     0x8a    /* RX buffer read pointer */
+#define ERXWRPT     0x8c    /* RX buffer write pointer */
+
+/** @} */
```
newline...

haukepetersen commented on drivers/encx24j600/include/encx24j600_defines.h:
```diff
@@ -0,0 +1,204 @@
+
+#define ERXRDPT     0x8a    /* RX buffer read pointer */
+#define ERXWRPT     0x8c    /* RX buffer write pointer */
+
```
newline...

haukepetersen commented on drivers/encx24j600/include/encx24j600_defines.h:
```diff
@@ -0,0 +1,204 @@
+#define PHCON2  0x11
+#define PHSTAT2 0x1b
+#define PHSTAT3 0x1f
+
```
newline ff.

haukepetersen commented on drivers/encx24j600/include/encx24j600_internal.h:
```diff
@@ -0,0 +1,44 @@
+
+    /* Receive Status Vector */
+    uint16_t frame_len;         /**< lenght of ethernet frame including 4 bytes
+                                  checksum */
```
indention is of

OlegHahm replied:
seems still to be off

haukepetersen commented on drivers/encx24j600/Makefile:
```diff
@@ -0,0 +1,3 @@
+MODULE = encx24j600
```
if I learned it correctly, you can remove this line as the module name will be automatically set

haukepetersen commented:
There seem to be some development leftovers in the code. Could you clean them up and rebase to master (to sync with the newest GPIO interface changes)? Thx.

jfischer-phytec-iot commented:
@kaspar030 
I have tested your driver and it seems to work. However, it has problems when more than one packet arrives at once and needs more love. I find it a pity that you can not use the hardware CS without additional GPIO (see kw2xrf driver). I'll add a couple of comments today.

kaspar030 commented:
On 06/16/15 09:26, Johann Fischer wrote:
> it has problems when more than one packet arrives at once

Could you elaborate?

jfischer-phytec-iot commented:
@kaspar030 
I will describe my test setup tonight.

jfischer-phytec-iot commented on drivers/encx24j600/encx24j600.c:
```diff
@@ -0,0 +1,424 @@
+static void cmdn(encx24j600_t *dev, uint8_t cmd, char *out, char *in, int len) {
+    gpio_clear(dev->cs);
+    spi_transfer_byte(dev->spi, cmd, NULL);
+    spi_transfer_bytes(dev->spi, out, in, len);
```
Please replase it (L168, L169) with spi_transfer_regs(dev->spi, cmd, out, in, len);

jfischer-phytec-iot commented on drivers/encx24j600/encx24j600.c:
```diff
@@ -0,0 +1,424 @@
+    dev->rx_next_ptr = hdr.rx_next_ptr;
+
+    reg_set(dev, ERXTAIL, dev->rx_next_ptr - 2);
+
```
You should not just go out of the function here.
It can be even more packets in the FIFO. The interrupt is configured for falling edge and will remain low as long as packets are available. If more than 1 packet arrives at a time then nothing works more. So I understood it. I have tested it with trafgen(netsniff-ng) and it has been confirmed.

kaspar030 replied:
I'll investigate. Could you try changing the 'if' to a 'while' in "receive_packets' (l350+)?

jfischer-phytec-iot replied:
``` 
trafgen --dev eth0 --conf trafgen.cfg -n 5

2015-06-17 23:09:04,331 - INFO # encx24j600: starting initialization...
2015-06-17 23:09:04,536 - INFO # encx24j600: initialization complete.
2015-06-17 23:09:04,539 - INFO # dev_eth: mac:  0:1e:c0:e3:73:b4
2015-06-17 23:09:04,542 - INFO # dev_eth: dev=0x1fffd0b0 link DOWN
2015-06-17 23:09:06,164 - INFO # main: woke up
2015-06-17 23:09:06,166 - INFO # dev_eth: dev=0x1fffd0b0 link UP
2015-06-17 23:09:19,478 - INFO # main: woke up
2015-06-17 23:09:19,482 - INFO # encx24j600: erxhead=0x5388 PKTCNT=1
2015-06-17 23:09:19,484 - INFO # dev_eth_rx_handler dev=0x1fffd0b0
2015-06-17 23:09:19,489 - INFO # encx24j600: ERXTAIL=0x5ffe frame_len=64 newnext=0x5388
2015-06-17 23:09:19,493 - INFO # handle_incoming: received 60 bytes
2015-06-17 23:09:19,497 - INFO # encx24j600: erxhead=0x5388 PKTCNT=0
2015-06-17 23:09:35,198 - INFO # main: woke up
2015-06-17 23:09:35,202 - INFO # encx24j600: erxhead=0x54f0 PKTCNT=5
2015-06-17 23:09:35,205 - INFO # dev_eth_rx_handler dev=0x1fffd0b0
2015-06-17 23:09:35,209 - INFO # encx24j600: ERXTAIL=0x5386 frame_len=64 newnext=0x53d0
2015-06-17 23:09:35,213 - INFO # handle_incoming: received 60 bytes
2015-06-17 23:09:35,217 - INFO # encx24j600: erxhead=0x54f0 PKTCNT=4
2015-06-17 23:09:35,220 - INFO # dev_eth_rx_handler dev=0x1fffd0b0
2015-06-17 23:09:35,225 - INFO # encx24j600: ERXTAIL=0x53ce frame_len=64 newnext=0x5418
2015-06-17 23:09:35,228 - INFO # handle_incoming: received 60 bytes
2015-06-17 23:09:35,232 - INFO # encx24j600: erxhead=0x54f0 PKTCNT=3
2015-06-17 23:09:35,235 - INFO # dev_eth_rx_handler dev=0x1fffd0b0
2015-06-17 23:09:35,240 - INFO # encx24j600: ERXTAIL=0x5416 frame_len=64 newnext=0x5460
2015-06-17 23:09:35,244 - INFO # handle_incoming: received 60 bytes
2015-06-17 23:09:35,248 - INFO # encx24j600: erxhead=0x54f0 PKTCNT=2
2015-06-17 23:09:35,251 - INFO # dev_eth_rx_handler dev=0x1fffd0b0
2015-06-17 23:09:35,255 - INFO # encx24j600: ERXTAIL=0x545e frame_len=64 newnext=0x54a8
2015-06-17 23:09:35,259 - INFO # handle_incoming: received 60 bytes
2015-06-17 23:09:35,263 - INFO # encx24j600: erxhead=0x54f0 PKTCNT=1
2015-06-17 23:09:35,266 - INFO # dev_eth_rx_handler dev=0x1fffd0b0
2015-06-17 23:09:35,271 - INFO # encx24j600: ERXTAIL=0x54a6 frame_len=64 newnext=0x54f0
2015-06-17 23:09:35,274 - INFO # handle_incoming: received 60 bytes
2015-06-17 23:09:35,278 - INFO # encx24j600: erxhead=0x54f0 PKTCNT=0
```

looks better

OlegHahm commented:
@kaspar030, what's the state of this one?

kaspar030 commented:
@OlegHahm will update next week.

kaspar030 commented:
Ok, this is working for me know, and I think I've addressed all comments.
The PR contains some fixes to ng_netdev_eth, making it build for platforms other than native.

@jfischer-phytec-iot @haukepetersen Could you take another look?

kaspar030 commented:
- rebased

OlegHahm commented:
ACK - merge if Travis is fine.

kaspar030 commented:
- rebased

OlegHahm commented:
rebase required again

kaspar030 commented:
- rebased
@OlegHahm does your ACK hold?

OlegHahm commented:
Yes, please squash!

kaspar030 commented:
&go.

Dialine commented:
Trying this driver with board from MikroElektronika connected to Atmel Saml21-xpro.
gpio for SPI seems well initialized because during first step of initialization, 0x1234 is read from the device. But the loop waiting 0000 from EUDAST after System Reset never end!
With what board did you test the device from MikroElektronika ?
An idea why System Reset seems not work ?
